daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/SequenceParserBases.scala

```diff
@@ -0,0 +1,380 @@
+   *
+   * No backtracking supported.
+   */
+  protected def parseOneWithoutPoU(
+    parserArg: SequenceChildParser,
+    trd: TermRuntimeData,
+    pstate: PState): ParseAttemptStatus
+
+  /**
+   * Parses one iteration of an array/optional element, and returns
+   * a status indicating success/failure and the nature of that success/failure.
+   *
+   * Supports speculative parsing via backtracking.
+   */
+  protected def parseOneWithPoU(
+    parser: RepeatingChildParser,
+    erd: ElementRuntimeData,
+    pstate: PState,
+    priorState: PState.Mark,
+    ais: GoArrayIndexStatus,
+    isBounded: Boolean): ParseAttemptStatus
+
+  protected def zeroLengthSpecialChecks(pstate: PState, wasLastChildZeroLength: Boolean): Unit
+
+  final protected def checkN(pstate: PState): Boolean = {
+    if (pstate.arrayPos > pstate.tunable.maxOccursBounds) {
+      PE(pstate, "Occurs count %s exceeds implementation maximum of %s.", pstate.arrayPos, pstate.tunable.maxOccursBounds)
+      false
+    } else true
+  }
+
+  /**
+   * This parse method is used for both separated and unseparated sequences.
+   */
+  override protected def parse(pstate: PState): Unit = {
+    val children = childParsers
+
+    var scpIndex = 0
+    pstate.mpstate.groupIndexStack.push(1L) // one-based indexing
+
+    val limit = children.length
+
+    var wasLastChildZeroLength = false
+
+    //
+    // This loop iterates over the children terms of the sequence
+    //
+    while ((scpIndex < limit) && (pstate.processorStatus eq Success)) {
+      val child = children(scpIndex)
+      child match {
+        case parser: RepeatingChildParser => {
+          //
+          // The sequence child is an array/repeating element (or ooptional
+          // element as the runtime doesn't distinguish them.)
+          //
+          //
+          val min = parser.minRepeats(pstate)
+          val max = parser.maxRepeats(pstate)
+          val isBounded = parser.isBoundedMax(max)
+          val erd = parser.trd.asInstanceOf[ElementRuntimeData]
+
+          parser.startArray(pstate)
+
+          //
+          // There are two kinds of loops. Arrays which have points of uncertainty (PoU)
+          // where speculative parsing is used to determine how many occurrences,
+          // and specified-number of occurrences, where a number is known or is computed.
+          //
+          parser.hasPoU match {
+            case true => {
+              //
+              // This case for array/optionals where the number of occurences is
+              // determined by speculative parsing. OCK=implicit with min/maxOccurs
+              // different, or OCK=parsed.
+              //
+
+              //
+              // The beforeArrayState will be assigned the priorState before the whole array
+              // This is the same object as the priorState before the first
+              // occurrence.
+              //
+              var beforeArrayState: PState.Mark = null
```

Comment:
This is related to DAFFODIL-1960. In this issue, we are keeping a mark to the very beginning of an array and it is never discarded until the end of the array. That means we could potentially back track to where the array started so we cannot discard any data when streaming. This is common in formats like PCAP where the entire file is just a repetition of packets. Is there any way around this, so that beforeArrayState is discarded at some point? maybe via discriminators or something?